# Data Cleaning

This notebook contains the data cleaning and preprocessing steps
performed on the Online Retail II dataset.

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv("../data/raw/online_retail_II.csv")

In [3]:
df.shape

(1067371, 8)

In [4]:
df.isnull().sum()

Invoice             0
StockCode           0
Description      4382
Quantity            0
InvoiceDate         0
Price               0
Customer ID    243007
Country             0
dtype: int64

In [5]:
df.duplicated().sum()

np.int64(34335)

In [6]:
df = df.drop_duplicates()

In [7]:
df.shape

(1033036, 8)

In [8]:
df=df.dropna(subset=['Customer ID'])

In [9]:
df.shape

(797885, 8)

In [10]:
df["InvoiceDate"] = pd.to_datetime(df["InvoiceDate"])

In [11]:
df.dtypes

Invoice                   str
StockCode                 str
Description               str
Quantity                int64
InvoiceDate    datetime64[us]
Price                 float64
Customer ID           float64
Country                   str
dtype: object

In [12]:
(df["Quantity"] < 0).sum()

np.int64(18390)

In [13]:
df = df[df["Quantity"] > 0]

In [14]:
df.shape

(779495, 8)

In [15]:
(df["Price"] <= 0).sum()

np.int64(70)

In [16]:
df = df[df["Price"] > 0]

In [17]:
df.shape

(779425, 8)

In [18]:
df.isnull().sum()

Invoice        0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
Price          0
Customer ID    0
Country        0
dtype: int64

In [19]:
df.duplicated().sum()

np.int64(0)

In [20]:
df[["Quantity", "Price"]].describe()

,Quantity,Price
count,779425.000000,779425.000000
mean,13.489370,3.218488
std,145.855814,29.676140
min,1.000000,0.001000
25%,2.000000,1.250000
50%,6.000000,1.950000
75%,12.000000,3.750000
max,80995.000000,10953.500000


## Cleaning Summary

The following data cleaning steps were performed:

- Removed duplicate records.
- Removed records with missing Customer ID.
- Converted InvoiceDate from string to datetime.
- Removed transactions with negative Quantity, which represent returns/cancellations for this analysis.
- Removed transactions with zero or negative Price.
- Verified that no missing values remain.
- Verified that no duplicate records remain.
- Performed final validation of Quantity and Price.

After cleaning, the dataset contains **779,425 rows and 8 columns**.

The cleaned dataset is now ready for Exploratory Data Analysis (EDA).